File: greedy/MinMeetingRooms919_LintCode.java
/**
 * Definition of Interval:
 * public classs Interval {
 *     int start, end;
 *     Interval(int start, int end) {
 *         this.start = start;
 *         this.end = end;
 *     }
 * }
 */
public class MinMeetingRooms919_LintCode {
    /**
     * @param intervals: an array of meeting time intervals
     * @return: the minimum number of conference rooms required
     */
    //小根堆的思路
    public int minMeetingRooms(List<Interval> intervals) {
        // Write your code here
        Collections.sort(intervals,(i1,i2)->i1.start-i2.start);
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        pq.add(intervals.get(0).end);
        for (int i = 1; i < intervals.size(); i++) {
            //lintCode这里的case好像有问题，我开始忘了加=也过了
            //大于等于堆顶的当前最早结束时间，说明可以和堆顶公用一个会议室
            if (intervals.get(i).start >= pq.peek()) {
                pq.poll();
            }
            pq.add(intervals.get(i).end);
        }
        return pq.size();
    }
}

File: greedy/MinMeetingRooms919_LintCode.go
/**
 * Definition of Interval:
 * type Interval struct {
 *     Start, End int
 * }
 */

/**
 * @param intervals: an array of meeting time intervals
 * @return: the minimum number of conference rooms required
 */
import "sort"

type Pair struct {
    time  int
    isEnd bool
}

func minMeetingRooms(intervals []*Interval) int {
    var n = len(intervals)
    var list []*Pair
    var Max = func(a, b int) int {
        if a > b {
            return a
        }
        return b
    }
    for i := 0; i < n; i++ {
        list = append(list, &Pair{intervals[i].Start, false})
        list = append(list, &Pair{intervals[i].End, true})
    }
    sort.Slice(list, func(i int, j int) bool {
        return list[i].time < list[j].time
    })
    var res = 0
    var count = 0
    for _, p := range list {
        if p.isEnd {
            count--
        } else {
            count++
        }
        res = Max(count, res)
    }
    return res
}
